### PR TITLE
CI: Unpin MyPy

### PR DESCRIPTION
**Pull request by simonjayhawkins**

### CODE REVIEW

File: pandas/_config/config.py [hunk 1]
```diff
@@ -460,9 +460,7 @@ def register_option(
     path = key.split(".")
 
     for k in path:
-        # NOTE: tokenize.Name is not a public constant
-        # error: Module has no attribute "Name"  [attr-defined]
-        if not re.match("^" + tokenize.Name + "$", k):  # type: ignore[attr-defined]
+        if not re.match("^" + tokenize.Name + "$", k):
```
Comment (simonjayhawkins) [inline pandas/_config/config.py:463]: latest typeshed (bundled with mypy) exposes undocumented attributes

File: pandas/core/common.py [hunk 2]
```diff
@@ -277,6 +277,11 @@ def maybe_iterable_to_list(obj: Union[Iterable[T], T]) -> Union[Collection[T], T
     """
     if isinstance(obj, abc.Iterable) and not isinstance(obj, abc.Sized):
         return list(obj)
+    # error: Incompatible return value type (got
+    # "Union[pandas.core.common.<subclass of "Iterable" and "Sized">,
+    # pandas.core.common.<subclass of "Iterable" and "Sized">1, T]", expected
+    # "Union[Collection[T], T]")  [return-value]
+    obj = cast(Collection, obj)
```
Comment (simonjayhawkins) [inline pandas/core/common.py:284]: our signature here is not entirely correct. If we pass an Iterable with a \_\_len__ method which is not necessarily a collection, the object is returned unchanged.

I think best to leave signature unchanged for now and could use Protocol in the future.

File: pandas/core/computation/expr.py [hunk 3]
```diff
@@ -364,7 +364,7 @@ class BaseExprVisitor(ast.NodeVisitor):
 
     unary_ops = _unary_ops_syms
     unary_op_nodes = "UAdd", "USub", "Invert", "Not"
-    unary_op_nodes_map = dict(zip(unary_ops, unary_op_nodes))
+    unary_op_nodes_map = {k: v for k, v in zip(unary_ops, unary_op_nodes)}
```
Comment (simonjayhawkins) [inline pandas/core/computation/expr.py:367]: mypy giving 

error: Need type annotation for 'unary_op_nodes_map' (hint: "unary_op_nodes_map: Dict[<type>, <type>] = ...")  [var-annotated]

mypy is ok with a dictcomp, or could add Dict[str, str] as variable type annotation instead

File: pandas/core/indexes/frozen.py [hunk 4]
```diff
@@ -103,5 +103,7 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f"{type(self).__name__}({str(self)})"
 
-    __setitem__ = __setslice__ = __delitem__ = __delslice__ = _disabled
-    pop = append = extend = remove = sort = insert = _disabled
+    __setitem__ = __setslice__ = _disabled  # type: ignore[assignment]
+    __delitem__ = __delslice__ = _disabled  # type: ignore[assignment]
+    pop = append = extend = _disabled  # type: ignore[assignment]
+    remove = sort = insert = _disabled  # type: ignore[assignment]
```
Comment (simonjayhawkins) [inline pandas/core/indexes/frozen.py:109]: we probably could make FrozenList Generic to allow type annotations such as FrozenList[Label]

This is an exercise for another day.

File: setup.cfg [hunk 5]
```diff
@@ -127,10 +127,7 @@ show_error_codes = True
 [mypy-pandas.tests.*]
 check_untyped_defs=False
 
-[mypy-pandas.conftest]
-ignore_errors=True
-
-[mypy-pandas.tests.tools.test_to_datetime]
```
Comment (simonjayhawkins) [inline setup.cfg:133]: errors were fixed with latest typeshed

File: setup.cfg [hunk 6]
```diff
@@ -127,10 +127,7 @@ show_error_codes = True
...
+[mypy-pandas.conftest,pandas.tests.window.conftest]
```
Comment (simonjayhawkins) [inline setup.cfg:130]: similar error in pandas.tests.window.conftest as in pandas.conftest,pandas, so bundled together. 

this is a mypy inference issue we have encountered before

error: List item 0 has incompatible type "ParameterSet"; expected "Sequence[Collection[object]]"  [list-item]

File: pandas/core/common.py (hunk 2 above)
Comment (simonjayhawkins) [reply pandas/core/common.py:284]: since i'm lying to the type checker a #type: ignore is probably more preferable to a cast.

File: setup.cfg [hunk 7]
```diff
@@ -226,6 +277,9 @@ check_untyped_defs=False
 [mypy-pandas.io.excel._util]
 check_untyped_defs=False
 
+[mypy-pandas.io.excel._xlsxwriter]
+check_untyped_defs=False
+
```
Comment (simonjayhawkins) [inline setup.cfg:282]: errors in this module fixed in #35995

File: pandas/_config/config.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_config/config.py:463]: does this mean theres a lower bound for what mypy we want?

**Comment by jbrockmendel** [comment]
LGTM

File: pandas/_config/config.py (hunk 1 above)
Comment (simonjayhawkins) [reply pandas/_config/config.py:463]: I think that in general it is recommended to pin mypy see https://mypy.readthedocs.io/en/stable/existing_code.html?highlight=pin#continuous-integration

> does this mean theres a lower bound for what mypy we want?

with warn_unused_ignores, yes.

Comment (simonjayhawkins) [reply pandas/_config/config.py:463]: so I could change to mypy=0.782 (best practice) or mypy>=0.782 (see https://github.com/pandas-dev/pandas/issues/27568#issuecomment-528938502)

**Comment by jbrockmendel** [comment]
LGTM

File: environment.yml [hunk 8]
```diff
@@ -21,7 +21,7 @@ dependencies:
   - flake8-comprehensions>=3.1.0  # used by flake8, linting of unnecessary comprehensions
   - flake8-rst>=0.6.0,<=0.7.0  # linting of code blocks in rst files
   - isort>=5.2.1  # check that imports are in the right order
-  - mypy=0.730
+  - mypy
```
Comment (jreback) [inline environment.yml:24]: would prefer we pin exactly for this, or at the very least a lower bound.

File: pandas/_config/config.py (hunk 1 above)
Comment (jreback) [reply pandas/_config/config.py:463]: commented above, prefer either of these to no version

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by WillAyd** [comment]
lgtm ex @jreback comments; definitely need to get this updated given all of the changes between 0.73 and now

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @simonjayhawkins 

note may want to add this to the whatsnew for development changes (we usually list pytest version changes)

**Pull request merged**